rest/client/iter.go

```diff
@@ -0,0 +1,54 @@
+package client
+
+// Iter defines an iterator type that should be used when implementing list methods. It's intended to
+// be embedded in a domain specific iterator struct.
+type Iter struct {
+	query Query
+
+	page    ListResponse
+	current interface{}
+	results []interface{}
+
+	err error
+}
+
+// Query defines a closure that domain specific iterators must implement. The implementation should
+// include a call to the API and should return the API response with a separate slice of the results.
+type Query func(string) (ListResponse, []interface{}, error)
+
+// GetIter returns a new initialized iterator. This method automatically makes the first query to
+// populate the results. List methods should use this helper method when building domain specific
+// iterators.
+func GetIter(url string, query Query) *Iter {
+	it := &Iter{query: query}
+	it.page, it.results, it.err = it.query(url)
+	return it
+}
+
+// Next moves the iterator to the next result.
+func (it *Iter) Next() bool {
+	if len(it.results) == 0 && it.page.NextPageURL() != "" {
+		it.page, it.results, it.err = it.query(it.page.NextPageURL())
+	}
+	if it.err != nil || len(it.results) == 0 {
+		return false
+	}
+	it.current = it.results[0]
+	it.results = it.results[1:]
+	return true
+}
+
+// Page returns the current results page in the form of an API response.
+func (it *Iter) Page() ListResponse {
+	return it.page
+}
+
+// Current returns the result that the iterator is currently pointing to.
+func (it *Iter) Current() interface{} {
```

Comment:
I'm not sure there's a consistent naming preference out there. I've seen `Item()` or `Value()`. I think those might be clearer than `Current()`.

FWIW, this Iter and the desire to make a "common" one is probably ripe for generics use. We side step this internally by generating code that is type specific. I think we should either have generated type specific code here or use generics.

Reply:
stripe uses `Current()` but i think i'll change it to `Value()`. also, i generally agree that generics could make sense here but i don't think we're trying to use that feature heavily just yet

Reply:
Now that I've seen how it's used in Quotes I think my concerns about it being held wrong are resolved.